README.md
<p class="josman-to-strip" align="center">
<img alt="DiverCLI" src="docs/img/diversicon-core-writing-100px.png" >
<br/>
</p>

#### About

Quick and dirty Java command line app to generate lexicons. Part of [Diversicon tools](http://diversicon-kb.eu/tools#java-libraries).

|**Usage**|**License**|**Roadmap**|**Contributing**|
|-----------|---------|-----------|----------------|
| See [docs](docs) |[GPL](LICENSE.txt) | See [project milestones](../../milestones) | See [wiki](../../wiki)|


**Features**

- performs basic conversion from Princeton Wordnet 3.1 to LMF XML format
- improves Wordnet id generation
- adds WordForms for lemma variations in WordNet 


#### Dependencies

Works with a modified DKPRO UBY 0.7.0, using wordnet component at 0.7.0-div-SNAPSHOT in the branch `develop-div`: https://github.com/diversicon-kb/dkpro-uby/blob/develop-div/de.tudarmstadt.ukp.uby.integration.wordnet-gpl


#### Credits

Design:

* Fiona McNeill - Heriot-Watt University, Edinburgh - f.mcneill at hw.ac.uk 
* Gabor Bella - DISI, University of Trento -  gabor.bella at unitn.it
* David Leoni - Heriot Watt University, Edinburgh - david.leoni at unitn.it

Programming:  

* David Leoni - Heriot Watt University, Edinburgh - david.leoni at unitn.it

Based on <a href="http://dkpro.github.io/dkpro-uby/" target="blank">DKPRO UBY framework</a>, by UKP Lab, Technische Universität Darmstadt.

Made possible thanks to:

&emsp;&emsp;&emsp;<a href="https://www.hw.ac.uk/schools/mathematical-computer-sciences/departments/computer-science.htm" target="_blank"> <img src="docs/img/hw.webp" width="80px" style="vertical-align:middle;"> </a> &emsp;&emsp;&emsp;<a href="https://www.hw.ac.uk/schools/mathematical-computer-sciences/departments/computer-science.htm" target="_blank"> Heriot-Watt Computer Science Department </a>  

&emsp;<a href="http://kidf.eu" target="_blank"> <img style="vertical-align:middle;" width="140px" src="docs/img/kidf-scientia.png"> </a> &emsp; <a href="http://kidf.eu" target="_blank"> Knowledge in Diversity Foundation </a> <br/>


docs/CHANGES.md

DIVMAKER ${project.version} RELEASE NOTES
-----------------------------------

http://diversicon-kb.eu/divmaker  


### 0.1.0

TODO Date  
 
Closed issues: https://github.com/diversicon-kb/divmaker/milestone/1?closed=1

- conversion from Princeton Wordnet 3.1 to LMF XML
- works with a modified UBY 0.7.0, using version wordnet at 0.7.0-div-SNAPSHOT in the branch `develop-div`: 
https://github.com/diversicon-kb/dkpro-uby/blob/develop-div/de.tudarmstadt.ukp.uby.integration.wordnet-gpl


src/test/java/eu/kidf/diversicon/maker/test/Tests.java
package eu.kidf.diversicon.maker.test;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tukaani.xz.XZInputStream;

public class Tests {
    
    private static final Logger LOG = LoggerFactory.getLogger(Tests.class);    
    
    /**
     * Expects a .xml file. If not found, a file .xml.xz
     * is searched and decompressed in same folder and returned
     * 
     */
    public static File getDump(File file) {

        if (file.getAbsolutePath()
                .endsWith(".xml")
                || file.getAbsolutePath()
                       .endsWith(".h2.db")) {
            if (file.exists()) {
                return file;
            } else {

                File xz = new File(file.getAbsolutePath() + ".xz");

                if (!xz.exists()) {
                    throw new IllegalStateException("Couldn't find file " + xz.getAbsolutePath() + " to decompress!");
                }

                LOG.info("Decompressing " + xz.getAbsolutePath() + " ...");
                try {
                    FileInputStream fin = new FileInputStream(xz);
                    BufferedInputStream in = new BufferedInputStream(fin);
                    FileOutputStream out = new FileOutputStream(file);
                    XZInputStream xzIn = new XZInputStream(in);
                    final byte[] buffer = new byte[8192];
                    int n = 0;
                    while (-1 != (n = xzIn.read(buffer))) {
                        out.write(buffer, 0, n);
                    }
                    out.close();
                    xzIn.close();
                    LOG.info("Wrote " + file.getAbsolutePath() + " file.");
                    return file;
                } catch (Exception e) {
                    throw new RuntimeException("Error while decompressing " + file.getAbsolutePath(), e);
                }

            }

        } else {
            throw new IllegalArgumentException("Unsupported input file extension for file " + file.getAbsolutePath());
        }

    }
    
    public static File getDump(String str) {
        return getDump(new File(str));
    }
}


src/test/java/it/unitn/disi/diversicon/maker/test/Tests.java
package it.unitn.disi.diversicon.maker.test;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tukaani.xz.XZInputStream;

public class Tests {
    
    private static final Logger LOG = LoggerFactory.getLogger(Tests.class);    
    
    /**
     * Expects a .xml file. If not found, a file .xml.xz
     * is searched and decompressed in same folder and returned
     * 
     */
    public static File getDump(File file) {

        if (file.getAbsolutePath()
                .endsWith(".xml")
                || file.getAbsolutePath()
                       .endsWith(".h2.db")) {
            if (file.exists()) {
                return file;
            } else {

                File xz = new File(file.getAbsolutePath() + ".xz");

                if (!xz.exists()) {
                    throw new IllegalStateException("Couldn't find file " + xz.getAbsolutePath() + " to decompress!");
                }

                LOG.info("Decompressing " + xz.getAbsolutePath() + " ...");
                try {
                    FileInputStream fin = new FileInputStream(xz);
                    BufferedInputStream in = new BufferedInputStream(fin);
                    FileOutputStream out = new FileOutputStream(file);
                    XZInputStream xzIn = new XZInputStream(in);
                    final byte[] buffer = new byte[8192];
                    int n = 0;
                    while (-1 != (n = xzIn.read(buffer))) {
                        out.write(buffer, 0, n);
                    }
                    out.close();
                    xzIn.close();
                    LOG.info("Wrote " + file.getAbsolutePath() + " file.");
                    return file;
                } catch (Exception e) {
                    throw new RuntimeException("Error while decompressing " + file.getAbsolutePath(), e);
                }

            }

        } else {
            throw new IllegalArgumentException("Unsupported input file extension for file " + file.getAbsolutePath());
        }

    }
    
    public static File getDump(String str) {
        return getDump(new File(str));
    }
}


src/test/java/eu/kidf/diversicon/maker/test/DivMakerTest.java
package eu.kidf.diversicon.maker.test;

import static org.junit.Assert.assertTrue;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

import org.junit.BeforeClass;
import org.junit.Ignore;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import de.tudarmstadt.ukp.lmf.model.core.LexicalResource;
import de.tudarmstadt.ukp.lmf.transform.DBConfig;
import de.tudarmstadt.ukp.lmf.transform.LMFXmlWriter;
import de.tudarmstadt.ukp.lmf.transform.XMLToDBTransformer;
import eu.kidf.diversicon.core.DivConfig;
import eu.kidf.diversicon.core.Diversicon;
import eu.kidf.diversicon.core.Diversicons;

/**
 * 
 * TODO this class is full of useless stuff, do clean up!
 * 
 * @since 0.1.0
 */
public class DivMakerTest {

    private static final Logger LOG = LoggerFactory.getLogger(DivMakerTest.class);

    private static final String dtdPath = "ubyLmfDTD_1.0.dtd";
    private static final String dtdVersion = "1_0";

    private static final String DUMPS_DIVERSICON = "dumps/diversicon/";

    private static final String TARGET_DIV = "target/diversicon/";

    private static final String WN30_DIV = "div-wn30";

    @BeforeClass
    public static void beforeClass(){
        try {
            Files.createDirectories(Paths.get("target", "div"));            
        } catch (IOException e) {        
            throw new RuntimeException("Something went wrong!", e);
        }        
    }    

   



   
    public File lexicalResourceToXml(LexicalResource lexicalResource, File lmfXML) {
        try {
            LOG.info("Going to create XML...");

            LMFXmlWriter xmlWriter = new LMFXmlWriter(lmfXML.getAbsolutePath(), dtdPath);
            xmlWriter.writeElement(lexicalResource);
            xmlWriter.writeEndDocument();

            LOG.info("xml file saved: " + lmfXML.getAbsolutePath());

            return lmfXML;
        } catch (Exception ex) {
            throw new RuntimeException("Error while writing LMF XML!", ex);
        }
    }

    @Test
    @Ignore
    public void testXmlDumpToDb() {
        File xmlDump = Tests.getDump(DUMPS_DIVERSICON + WN30_DIV + ".xml");
        String outDb = TARGET_DIV + "wn30-from-dump";
        xmlToDb(xmlDump, outDb);

        // checkDb(outDb);
    }

    @Test
    @Ignore
    public void testDbToSql() {
        DBConfig dbConfig = Diversicons.h2InMemoryConfig("mydb", false);
        Diversicons.dropCreateTables(dbConfig);
        Diversicon div = Diversicon.connectToDb(DivConfig.of(dbConfig));
        div.importXml("dumps/uby/uby-wn30.xml.xz");
        String zipFilePath = TARGET_DIV + WN30_DIV + ".sql.zip";
        div.exportToSql(zipFilePath, true);
        assertTrue(new File(zipFilePath).exists());
        div.getSession().close();
    }

    
    @Test    
    @Ignore
    public void testConnectToDb() {
        DBConfig dbConfig = Diversicons.h2FileConfig("~/Da/prj/diversicon/dumps/div-wn31", true);
        Diversicon div = Diversicon.connectToDb(DivConfig.of(dbConfig));
                
        div.getSession().close();
        
        Diversicons.h2Execute("SHUTDOWN COMPACT", dbConfig);
    }
    
    
    @Test
    @Ignore
    public void testRestoreFileDb() throws IOException {
        Path tempDir = Files.createTempDirectory("divmaker-test");
        Diversicons.h2RestoreSql(DUMPS_DIVERSICON + WN30_DIV + ".zip",
                DivConfig.of(Diversicons.h2FileConfig(tempDir.toString() + "/temp-db", true)));
    }

    @Test
    @Ignore
    public void testRestoreZipToInMemoryDb() throws IOException {
        Path tempDir = Files.createTempDirectory("divmaker-test");
        Diversicons.h2RestoreSql(DUMPS_DIVERSICON + WN30_DIV + ".zip", 
                DivConfig.of(Diversicons.h2InMemoryConfig(tempDir.toString() + "/temp-db", 
                        false)));
    }
    
    @Test
    @Ignore
    public void testRestoreSqlToInMemoryDb() throws IOException {
        Path tempDir = Files.createTempDirectory("divmaker-test");
        Diversicons.h2RestoreSql(DUMPS_DIVERSICON + WN30_DIV + ".sql", 
                DivConfig.of(Diversicons.h2InMemoryConfig(tempDir.toString() + "/temp-db", false)));
    }

   
    public void checkDb(String outDbPath) {
        File outDb = new File(outDbPath + ".h2.db");
        assertTrue(outDb.exists());
        assertTrue(outDb.length() > 1000000);
    }    

    public void xmlToDb(File inputXml, String outDb) {

        try {
            if (!inputXml.exists()) {
                throw new RuntimeException("Input xml doesn't exist! Path is: " + inputXml.getAbsolutePath());
            }

            LOG.info("Going to populate H2 DB " + outDb + "...");

            DBConfig dbConfig = Diversicons.h2FileConfig(outDb, false);

            Diversicons.dropCreateTables(dbConfig);

            XMLToDBTransformer dbWriter = new XMLToDBTransformer(dbConfig);

            dbWriter.transform(inputXml, WN30_DIV);

            LOG.info("db saved: " + outDb);

        } catch (Exception ex) {
            throw new RuntimeException("Error while writing db!", ex);
        }
    }
}


src/main/java/eu/kidf/diversicon/maker/DivMaker.java
package eu.kidf.diversicon.maker;

import static eu.kidf.diversicon.core.internal.Internals.checkNotBlank;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import org.apache.commons.io.FilenameUtils;
import org.hibernate.SQLQuery;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tukaani.xz.LZMA2Options;
import org.tukaani.xz.UnsupportedOptionsException;
import org.tukaani.xz.XZOutputStream;

import de.tudarmstadt.ukp.lmf.model.core.LexicalResource;
import de.tudarmstadt.ukp.lmf.model.meta.MetaData;
import de.tudarmstadt.ukp.lmf.transform.DBConfig;
import de.tudarmstadt.ukp.lmf.transform.LMFXmlWriter;
import de.tudarmstadt.ukp.lmf.transform.wordnet.WNConverter;
import eu.kidf.diversicon.core.DivConfig;
import eu.kidf.diversicon.core.Diversicon;
import eu.kidf.diversicon.core.Diversicons;
import eu.kidf.diversicon.core.ImportConfig;
import eu.kidf.diversicon.core.LexResPackage;
import eu.kidf.diversicon.core.exceptions.DivException;
import eu.kidf.diversicon.core.internal.Internals;
import eu.kidf.diversicon.data.DivWn31;
import net.sf.extjwnl.JWNLException;
import net.sf.extjwnl.dictionary.Dictionary;

/**
 * Creates Diversicon databases in various formats ready for packaging.
 * 
 * Currently only processes Wordnet 3.1
 * 
 * @since 0.1.0
 */
public class DivMaker {

    private static final Logger LOG = LoggerFactory.getLogger(DivMaker.class);

    private static final String dtdPath = Diversicons.DTD_1_0_PUBLIC_URL;
    private static final String dtdVersion = "1.0";

    /**
     * @since 0.1.0
     */
    private static final String DUMPS_UBY = "dumps/uby/";

    /**
     * @since 0.1.0
     */
    private static final String DUMPS_DIV = "dumps/diversicon/";

    /**
     * version with dots, like '3.1'
     */
    private String version;

    LexResPackage pack;
    LexicalResource lexRes;
    List<String> errors;

    /**
     * @since 0.1.0
     */
    private DivMaker() {
        this.pack = new LexResPackage();
        this.errors = new ArrayList<>();
    }

    /**
     * Currently only processes Wordnet 3.1 and takes no arguments.
     * 
     * @since 0.1.0
     */
    public static void main(String args[]) {

        if (args.length > 0) {
            throw new DivException("Invalid arguments! Currently the program accepts no arguments at all!");
        }

        DivMaker m = new DivMaker();

        m.version = "3.1";
        m.pack = DivWn31.of();

        m.run();

    }

    /**
     * Version without dots, like '30'
     * 
     * @since 0.1.0
     */
    private String vNoDot() {
        return version.replace(".", "");
    }

    /**
     * @since 0.1.0
     */
    private void run() {

        LOG.info("");
        LOG.info("");
        LOG.info("****   Going to create:\n" + pack.toString());
        LOG.info("");
        LOG.info("");
        Internals.checkLexResPackage(pack);

        this.lexRes = wordnetToLexRes();
        

        File fDumps = new File(DUMPS_DIV);

        if (!fDumps.exists()) {
            LOG.info("Creating " + fDumps + "   ...");
            if (!fDumps.mkdirs()) {
                throw new DivException("Couldn't create directory " + fDumps);
            }

        }

        File xmlFile = new File(DUMPS_DIV + lexRes.getName() + ".xml");
        if (xmlFile.exists()) {
            xmlFile.delete();
        }

        File sqlFile = new File(DUMPS_DIV + lexRes.getName() + ".sql");
        if (sqlFile.exists()) {
            sqlFile.delete();
        }

        File h2dbFile = new File(DUMPS_DIV + lexRes.getName() + ".h2.db");
        if (h2dbFile.exists()) {
            h2dbFile.delete();
        }

        LOG.info("****  Going to create XML to " + xmlFile.getAbsolutePath());
        LOG.info("****                   (make take several minutes...)");

        Diversicons.writeLexResToXml(lexRes, pack, xmlFile);
        LOG.info("****  Done creating " + xmlFile.getAbsolutePath());
        
        compress(xmlFile);        
        
        DBConfig dbConfig = Diversicons.h2FileConfig(DUMPS_DIV + lexRes.getName(), false);

        Diversicons.createTables(dbConfig);

        Diversicon div = Diversicon.connectToDb(DivConfig.of(dbConfig));
        ImportConfig config = new ImportConfig();
        config.setAuthor("David Leoni");
        config.setDescription("Import for making Wordnet 3.1 Diversicon distribution");
        config.setFileUrls(Internals.newArrayList(xmlFile.getAbsolutePath()));

        div.importFiles(config);

        try {
            LOG.info("****  Exporting to SQL " + sqlFile.getAbsolutePath());            
            div.exportToSql(sqlFile.getAbsolutePath(), false);
            compress(sqlFile);
            
        } catch (Exception ex) {
            error("FAILED EXPORTING TO SQL!", ex);
        }
        
        
        try {
            LOG.info("****  Compacting db");
            Diversicons.h2Execute("SHUTDOWN COMPACT", dbConfig);
            LOG.info("****  Done compacting db.");
            if (!div.getSession()
                    .isOpen()) {
                div.getSession()
                   .close();
            }

            compress(h2dbFile);
        } catch (Exception ex) {
            error("FAILED COMPACTING THE DB!", ex);
        }


        if (errors.size() > 0) {
            LOG.error("");
            LOG.error("");
            LOG.error("!!!!  DONE WITH ERRORS (look at log for more info):");
            for (String err : errors) {
                LOG.error(" - " + err);
            }
            LOG.info("");
        } else {
            LOG.info("");
            LOG.info("");
            LOG.info("****  DONE!!!   ALLELUJA!!!  ");
            LOG.info("");
            LOG.info("NOTE: probably the db somehow was not closed properly, ");            
            LOG.info("For more info about this, see https://github.com/diversicon-kb/divmaker/issues/2");
        }

    }

    private void error(String msg, Exception ex) {
        errors.add(msg);
        LOG.error("");
        LOG.error("!!!!  " + msg, ex);
        LOG.error("");
    }

    /**
     * Returns a new LexicalResource
     * 
     * @since 0.1.0
     */
    public LexicalResource wordnetToLexRes() {
        checkNotBlank(version, "Invalid version!");

        File wordnetOnDisk = new File("wn" + vNoDot() + "/");

        Internals.checkArgument(wordnetOnDisk.exists(),
                "I need a wordnet distribution at " + wordnetOnDisk.getAbsolutePath());

        Dictionary dictionary;
        try {
            dictionary = Dictionary.getResourceInstance(
                    "/net/sf/extjwnl/data/wordnet/wn" + vNoDot() + "/res_properties.xml");
        } catch (JWNLException e) {
            throw new DivException("I need a wordnet distribution as maven package!", e);
        }

        if (dictionary == null) {
            throw new DivException("no dictionary!");
        }

        LexicalResource lexRes = new LexicalResource();

        lexRes.setName(DivWn31.NAME);
        lexRes.setDtdVersion(dtdVersion);

        WNConverter c = new WNConverter(
                "wn" + vNoDot(),
                wordnetOnDisk,
                dictionary,
                lexRes,
                version,
                dtdVersion);

        c.toLMF();

        LOG.info("Created lexical resource in memory!");

        return lexRes;

    }
    
    /**
     * This function <b>NEVER</b> throws
     * @since 0.1.0
     */
    private void compress(File src){
        
        String path;
        if (src == null){
            error("FOUND NULL PATH FOR COMPRESSING !", new Exception(""));
            return;
        } else {
            path = src.getAbsolutePath();
        }
        
        try {            
            compressXz(src);
        } catch (Exception ex){
            error("FAILED COMPRESSING XZ OF " + path, ex);
        }
        
        try {            
            compressZip(src);
        } catch (Exception ex){
            error("FAILED COMPRESSING ZIP OF " + path, ex);
        }
    }

    /**
     * Creates {@code src.xz} file in same {@code src} directory
     * 
     * @since 0.1.0
     */
    private static void compressXz(File src) {
        Internals.checkNotNull(src);
                               
        String dst = src.getAbsolutePath() +  ".xz";
        
        LOG.info("");
        LOG.info("**** Creating " + src.getAbsolutePath() + ".xz   ....");
        
        try (FileInputStream inFile = new FileInputStream(src)){
                        
            FileOutputStream outfile = new FileOutputStream(dst);

            LZMA2Options options = new LZMA2Options();
            
            options.setPreset(7);

            try (XZOutputStream out = new XZOutputStream(outfile, options)){
                byte[] buf = new byte[8192];
                int size;
                while ((size = inFile.read(buf)) != -1)
                    out.write(buf, 0, size);
                out.finish();                
            }

            LOG.info("**** Created " + src.getAbsolutePath() + ".xz");
            
        } catch (IOException e) {
            throw new RuntimeException(e);        
        }

    }

    /**
     * Creates {@code src.zip} file in same {@code src} directory
     * 
     * @since 0.1.0
     */    
    public void compressZip(File src){
        Internals.checkNotNull(src);
        
        String dst = src.getAbsolutePath() +  ".zip";
        
        LOG.info("");
        LOG.info("**** Creating " + src.getAbsolutePath() + ".zip   ....");
        
        try {
                        
            StringBuilder sb = new StringBuilder();
            sb.append("Test String");
            
            ZipOutputStream zos = new ZipOutputStream(new FileOutputStream(dst));
            ZipEntry e = new ZipEntry(src.getName());
            zos.putNextEntry(e);
            
            FileInputStream in = new FileInputStream(src);
            
            byte[] buffer = new byte[1024];
            int len;
            while ((len = in.read(buffer)) > 0) {
                zos.write(buffer, 0, len);
            }
           
            zos.closeEntry();
            zos.close();                                   

            LOG.info("**** Created " + src.getAbsolutePath() + ".xz");
            
        } catch (IOException e) {
            throw new RuntimeException(e);        
        }
        
        
    }

}
